acceptable/mocks.py

```diff
@@ -0,0 +1,461 @@
+
+    def attach_callback(self, methods, url, callback):
+        for method in methods:
+            responses.mock.add_callback(method, url, callback)
+        self._attach()
+
+    def detach_callback(self, methods, url, callback):
+        for method in methods:
+            responses.mock.remove(method, url)
+        self._detach()
+
+
+_responses_manager = ResponsesManager()
+
+
+class responses_mock_context(object):
+    """A context manager to allow the safe (single threaded) use of the global
+     responses mock.
+
+     The object returned by this context manager is the responses.mock 
+     RequestsMock object.
+
+     You should us this context manager rather than any other instance of
+     RequestsMock.
+    """
+
+    def __enter__(self):
+        _responses_manager._attach()
+        return responses.mock
+
+    def __exit__(self, *args):
+        _responses_manager._detach()
+
+
+Call = namedtuple("Call", "request response error".split())
+
+
+class CallRecorder(object):
+    def __init__(self):
+        self._calls = []
+
+    def record(self, mock, request, response, error):
+        self._calls.append((mock, Call(request, response, error)))
+
+    def get_calls(self):
+        return [c for m, c in self._calls]
+
+    def get_calls_for(self, mock):
+        return [c for m, c in self._calls if m == mock]
+
+    def get_calls_for_matching(self, mock, pattern):
+        if not hasattr(pattern, "search"):
+            pattern = re.compile(pattern)
+        return [c for c in self.get_calls_for(mock) if pattern.search(c.request.url)]
+
+    def get_calls_matching(self, pattern):
+        if not hasattr(pattern, "search"):
+            pattern = re.compile(pattern)
+        return [c for m, c in self._calls if pattern.search(c.request.url)]
+
+
+EndpointSpec = namedtuple(
+    "EndpointSpec", ["name", "location", "methods", "request_schema", "response_schema"]
+)
+
+
+VALIDATION_ERROR_TEXT = """
+{}:
+{!r}
+
+did not match schema:
+{!r}
+
+for service {!r} endpoint {!r} on url {!r} errors where:
+{!r}
+"""
+
+
+class EndpointMock(object):
+    def __init__(
+        self,
+        call_recorder,
+        service_name,
+        name,
+        methods,
+        url,
+        request_schema,
+        response_schema,
+        response_callback,
+    ):
+        self._call_recorder = call_recorder
+        self._service_name = service_name
+        self._name = name
+        self._methods = methods
+        self._url = url
+        self._request_schema = request_schema
+        self._response_schema = response_schema
+        self._response_callback = response_callback
+
+    @property
+    def service_name(self):
+        return self._service_name
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def url(self):
+        return self._url
+
+    @property
+    def methods(self):
+        return list(self._methods)
+
+    @property
+    def call_recorder(self):
+        return self._call_recorder
```

Comment:
I don't think that properties are worth it here - why not just use regular members?

Reply:
These are properties that should be immutable but are still useful to read. I think its useful to avoid misunderstanding.